File: python/ga.py
import numpy as np
import torch
from snake_env import SnakeEnv
from network import forward_batch, random_population, PARAM_COUNT


def evaluate_population(
    params: torch.Tensor,
    grid_size: int,
    device: torch.device,
) -> np.ndarray:
    """
    Run all N snake episodes in parallel using batched GPU inference.
    All snakes share the same food sequence so fitness scores are comparable.
    """
    N = params.shape[0]
    max_dist = (grid_size - 1) * 2  # max possible manhattan distance on grid

    food_seq = [
        (int(np.random.randint(0, grid_size)), int(np.random.randint(0, grid_size)))
        for _ in range(500)
    ]
    envs = [SnakeEnv(grid_size, food_sequence=food_seq) for _ in range(N)]

    obs_np = np.stack([env.reset() for env in envs])  # (N, 18)
    alive  = np.ones(N, dtype=bool)

    global_max = grid_size * grid_size * 10

    for _ in range(global_max):
        if not alive.any():
            break

        obs_t = torch.from_numpy(obs_np).to(device)
        with torch.no_grad():
            actions = forward_batch(params, obs_t).argmax(dim=1).cpu().numpy()

        for i in range(N):
            if not alive[i]:
                continue
            obs_np[i], _, done = envs[i].step(int(actions[i]))
            if done:
                alive[i] = False

    fitness = np.zeros(N, dtype=np.float32)
    for i, env in enumerate(envs):
        hr, hc = env.snake[0]
        fr, fc = env.food
        dist_bonus = max_dist - (abs(hr - fr) + abs(hc - fc))
        if env.score > 0:
            fitness[i] = env.score * 1000 - env.steps * 0.1 + dist_bonus
        else:
            fitness[i] = float(dist_bonus)

    return fitness


class GeneticAlgorithm:
    def __init__(
        self,
        population_size: int = 500,
        grid_size: int = 20,
        device: torch.device = torch.device("cpu"),
        elite_frac: float = 0.1,
        selection_pressure: float = 3.0,
        mutation_std: float = 0.3,
        mutation_rate: float = 0.05,
        crossover_rate: float = 0.5,
    ):
        self.N                  = population_size
        self.grid_size          = grid_size
        self.device             = device
        self.elite_n            = max(1, int(population_size * elite_frac))
        self.selection_pressure = selection_pressure
        self.mutation_std       = mutation_std
        self.mutation_rate      = mutation_rate
        self.crossover_rate     = crossover_rate

        self._rank_probs: np.ndarray = self._build_rank_probs(population_size, selection_pressure)

        self.params   = random_population(population_size, device)
        self.fitness  = np.zeros(population_size, dtype=np.float32)
        self.generation = 0
        self._ranked_params: list = []

    def step(self) -> dict:
        """Run one generation: evaluate → select → reproduce."""
        self.fitness = evaluate_population(self.params, self.grid_size, self.device)
        self.generation += 1

        ranked = np.argsort(self.fitness)[::-1]
        self._ranked_params = [self.params[int(idx)].clone() for idx in ranked]

        new_params = self.params.clone()

        for i, idx in enumerate(ranked[: self.elite_n]):
            new_params[i] = self.params[idx]

        for i in range(self.elite_n, self.N):
            if np.random.rand() < self.crossover_rate:
                p1 = self.params[self._rank_select(ranked)]
                p2 = self.params[self._rank_select(ranked)]
                child = self._crossover(p1, p2)
            else:
                child = self.params[self._rank_select(ranked)].clone()
            self._mutate(child)
            new_params[i] = child

        self.params = new_params

        return {
            "generation":    self.generation,
            "best_fitness":  float(self.fitness[ranked[0]]),
            "avg_fitness":   float(self.fitness.mean()),
            "best_score":    int(self.fitness[ranked[0]]) // 1000,
        }

    @staticmethod
    def _build_rank_probs(n: int, selection_pressure: float) -> np.ndarray:
        weights = np.linspace(selection_pressure, 1.0, n)
        return weights / weights.sum()

    def _rank_select(self, ranked: np.ndarray) -> int:
        rank_pos = np.random.choice(self.N, p=self._rank_probs)
        return int(ranked[rank_pos])

    def _crossover(self, p1: torch.Tensor, p2: torch.Tensor) -> torch.Tensor:
        mask = torch.rand(PARAM_COUNT, device=self.device) < 0.5
        return torch.where(mask, p1, p2)

    def _mutate(self, individual: torch.Tensor):
        mask_fine = torch.rand(PARAM_COUNT, device=self.device) < self.mutation_rate
        noise_fine = torch.randn(PARAM_COUNT, device=self.device) * self.mutation_std
        mask_large = torch.rand(PARAM_COUNT, device=self.device) < 0.02
        noise_large = torch.randn(PARAM_COUNT, device=self.device) * (self.mutation_std * 5)
        individual.add_(mask_fine * noise_fine + mask_large * noise_large)

    def top_n_params(self, n: int) -> list:
        return self._ranked_params[:n]

File: rust/src/ga.rs
use rand::Rng;
use rand_distr::{Distribution, Normal};
use rayon::prelude::*;

use crate::gpu::GpuInference;
use crate::network::{forward, argmax, random_population, IN_SIZE};
use crate::snake_env::SnakeEnv;

pub struct Stats {
    pub generation:   usize,
    pub best_fitness: f32,
    pub avg_fitness:  f32,
    pub best_score:   usize,
}

pub struct GeneticAlgorithm {
    pub params:          Vec<Vec<f32>>,
    pub fitness:         Vec<f32>,
    pub generation:      usize,
    pub ranked_params:   Vec<Vec<f32>>,

    pub population_size: usize,
    pub grid_size:       i32,
    elite_n:             usize,
    mutation_std:        f32,
    mutation_rate:       f32,
    crossover_rate:      f32,
    rank_probs:          Vec<f64>,
    gpu:                 Option<GpuInference>,
}

impl GeneticAlgorithm {
    pub fn new(
        population_size:    usize,
        grid_size:          i32,
        elite_frac:         f32,
        selection_pressure: f32,
        mutation_std:       f32,
        mutation_rate:      f32,
        crossover_rate:     f32,
        gpu:                Option<GpuInference>,
    ) -> Self {
        let elite_n    = ((population_size as f32 * elite_frac) as usize).max(1);
        let rank_probs = build_rank_probs(population_size, selection_pressure);

        Self {
            params:             random_population(population_size),
            fitness:            vec![0.0; population_size],
            generation:         0,
            ranked_params:      vec![],
            population_size,
            grid_size,
            elite_n,
            mutation_std,
            mutation_rate,
            crossover_rate,
            rank_probs,
            gpu,
        }
    }

    pub fn step(&mut self) -> Stats {
        self.fitness = match &self.gpu {
            Some(g) => Self::evaluate_gpu(g, &self.params, self.grid_size),
            None    => Self::evaluate_cpu(&self.params, self.grid_size),
        };
        self.generation += 1;

        // Sort indices best → worst
        let mut ranked: Vec<usize> = (0..self.population_size).collect();
        ranked.sort_unstable_by(|&a, &b| {
            self.fitness[b].partial_cmp(&self.fitness[a]).unwrap()
        });

        self.ranked_params = ranked.iter().map(|&i| self.params[i].clone()).collect();

        let mut new_params = self.params.clone();

        // Elites
        for (slot, &src) in ranked[..self.elite_n].iter().enumerate() {
            new_params[slot] = self.params[src].clone();
        }

        // Offspring
        let mut rng = rand::thread_rng();
        for i in self.elite_n..self.population_size {
            let child = if rng.gen::<f32>() < self.crossover_rate {
                let p1 = &self.params[self.rank_select(&ranked, &mut rng)];
                let p2 = &self.params[self.rank_select(&ranked, &mut rng)];
                crossover(p1, p2, &mut rng)
            } else {
                self.params[self.rank_select(&ranked, &mut rng)].clone()
            };
            new_params[i] = self.mutate(child, &mut rng);
        }

        self.params = new_params;

        let best_fitness = self.fitness[ranked[0]];
        Stats {
            generation:   self.generation,
            best_fitness,
            avg_fitness:  self.fitness.iter().sum::<f32>() / self.population_size as f32,
            best_score:   (best_fitness as usize) / 1000,
        }
    }

    pub fn top_n_params(&self, n: usize) -> Vec<Vec<f32>> {
        self.ranked_params[..n.min(self.ranked_params.len())].to_vec()
    }

    // ── Private ──────────────────────────────────────────────────────────────

    fn food_seq(grid_size: i32) -> Vec<(i32, i32)> {
        let mut rng = rand::thread_rng();
        (0..500).map(|_| (rng.gen_range(0..grid_size), rng.gen_range(0..grid_size))).collect()
    }

    fn fitness_of(env: &SnakeEnv, max_dist: f32) -> f32 {
        let (hr, hc) = env.snake[0];
        let (fr, fc) = env.food;
        let dist_bonus = max_dist - ((hr - fr).abs() + (hc - fc).abs()) as f32;
        if env.score > 0 {
            env.score as f32 * 1000.0 - env.steps as f32 * 0.1 + dist_bonus
        } else {
            dist_bonus
        }
    }

    /// GPU path: sequential game loop, GPU batch inference each step.
    fn evaluate_gpu(gpu: &GpuInference, params: &[Vec<f32>], grid_size: i32) -> Vec<f32> {
        let n        = params.len();
        let max_dist = ((grid_size - 1) * 2) as f32;
        let food_seq = Self::food_seq(grid_size);

        let mut envs: Vec<SnakeEnv> = (0..n)
            .map(|_| SnakeEnv::new(grid_size, food_seq.clone()))
            .collect();

        // Flat obs buffer: n × IN_SIZE, updated in-place each step.
        let mut obs_flat: Vec<f32> = envs.iter_mut()
            .flat_map(|e| e.reset())
            .collect();
        let mut alive = vec![true; n];

        gpu.upload_params(params);

        loop {
            if alive.iter().all(|&a| !a) { break; }

            let actions = gpu.infer(&obs_flat);

            for i in 0..n {
                if !alive[i] { continue; }
                let (new_obs, done) = envs[i].step(actions[i]);
                let base = i * IN_SIZE;
                obs_flat[base..base + IN_SIZE].copy_from_slice(&new_obs);
                if done { alive[i] = false; }
            }
        }

        envs.iter().map(|e| Self::fitness_of(e, max_dist)).collect()
    }

    /// CPU path: rayon parallel, one thread per individual.
    fn evaluate_cpu(params: &[Vec<f32>], grid_size: i32) -> Vec<f32> {
        let max_dist = ((grid_size - 1) * 2) as f32;
        let food_seq = Self::food_seq(grid_size);

        params.par_iter().map(|p| {
            let mut env = SnakeEnv::new(grid_size, food_seq.clone());
            let mut obs = env.reset();
            loop {
                let action = argmax(&forward(p, &obs));
                let (new_obs, done) = env.step(action);
                obs = new_obs;
                if done { break; }
            }
            Self::fitness_of(&env, max_dist)
        }).collect()
    }

    fn rank_select(&self, ranked: &[usize], rng: &mut impl Rng) -> usize {
        let r: f64 = rng.gen();
        let mut cum = 0.0f64;
        for (pos, p) in self.rank_probs.iter().enumerate() {
            cum += p;
            if r < cum {
                return ranked[pos];
            }
        }
        ranked[ranked.len() - 1]
    }

    fn mutate(&self, mut p: Vec<f32>, rng: &mut impl Rng) -> Vec<f32> {
        let fine_dist  = Normal::new(0.0f32, self.mutation_std).unwrap();
        let large_dist = Normal::new(0.0f32, self.mutation_std * 5.0).unwrap();
        for v in p.iter_mut() {
            if rng.gen::<f32>() < self.mutation_rate {
                *v += fine_dist.sample(rng);
            }
            // Rare large jumps (2%) for occasional exploration
            if rng.gen::<f32>() < 0.02 {
                *v += large_dist.sample(rng);
            }
        }
        p
    }
}

fn build_rank_probs(n: usize, selection_pressure: f32) -> Vec<f64> {
    let sp = selection_pressure as f64;
    let weights: Vec<f64> = (0..n)
        .map(|i| sp - (sp - 1.0) * i as f64 / (n - 1).max(1) as f64)
        .collect();
    let total: f64 = weights.iter().sum();
    weights.iter().map(|w| w / total).collect()
}

fn crossover(p1: &[f32], p2: &[f32], rng: &mut impl Rng) -> Vec<f32> {
    p1.iter().zip(p2.iter()).map(|(&a, &b)| if rng.gen::<bool>() { a } else { b }).collect()
}
